Api refrence update tasks and keys (#1796)
meilisearch/documentation: reference/api/keys.md

[Learn more about managing keys and their rights](/learn/security/master_api_keys.md).

## Key object

```json
{
  "name": "Default Search API Key",
  "description": "Use it to search from the frontend code",
  "key": "0a6e572506c52ab0bd6195921575d23092b7f0c284ab4ac86d12346c33057f99",
  "uid": "74c9c733-3368-4738-bbe5-1d18a5fecb37",
  "actions": [
    "search"
  ],
  "indexes": [
    "*"
  ],
  "expiresAt": null,
  "createdAt": "2021-08-11T10:00:00Z",
  "updatedAt": "2021-08-11T10:00:00Z"
}
```

### `name`

**Type**: String
**Default value**: `null`
**Description**: A human-readable name for the key

### `description`

**Type**: String
**Default value**: `null`
**Description**: A description for the key. You can add any important information about the key here

### `uid`

**Type**: String
**Default value**: N/A
**Description**: A [uuid v4](https://www.sohamkamani.com/uuid-versions-explained) to identify the API key. If not specified, it is automatically generated by Meilisearch

### `key`

**Type**: String
**Default value**: N/A

> **Description**: An alphanumeric key value generated by Meilisearch by hashing the `uid` and the master key on API key creation. Used for authorization when [making calls to a protected Meilisearch instance](/learn/security/master_api_keys.md#communicating-with-a-protected-instance).

Reviewer: I'm not sure I like removing the period, but it felt inconsistent with the majority of the other field descriptions. Alternatively we could add periods to all other descriptions, but this solution seems more efficient.

```suggestion
**Description**: An alphanumeric key value generated by Meilisearch by hashing the `uid` and the master key on API key creation. Used for authorization when [making calls to a protected Meilisearch instance](/learn/security/master_api_keys.md#communicating-with-a-protected-instance)
```

This value is also used as the `{key}` path variable to [update](#update-a-key), [delete](#delete-a-key), or [get](#get-one-key) a specific key.

::: note
Since `key` is a hash of the `uid` and master key, `key` values are deterministic between instances sharing the same configuration. This means if the master key changes, all `key` values are automatically changed.

Since the `key` field depends on the master key, it is computed at runtime and therefore not propagated to dumps and snapshots. As a result, even if a malicious user comes into possession of your dumps or snapshots, they will not have access to your instance's API keys.
:::

### `actions`

**Type**: Array
**Default value**: N/A

> **Description**: An array of API actions permitted for the key, represented as strings. API actions are only possible on authorized [`indexes`](#indexes). `["*"]` for all actions.

Reviewer: ```suggestion
**Description**: An array of API actions permitted for the key, represented as strings. API actions are only possible on authorized [`indexes`](#indexes). `["*"]` for all actions
```

[Learn more about managing keys and their rights](/learn/security/master_api_keys.md).

## Key object

```json
{
  "name": "Default Search API Key",
  "description": "Use it to search from the frontend code",
  "key": "0a6e572506c52ab0bd6195921575d23092b7f0c284ab4ac86d12346c33057f99",
  "uid": "74c9c733-3368-4738-bbe5-1d18a5fecb37",
  "actions": [
    "search"
  ],
  "indexes": [
    "*"
  ],
  "expiresAt": null,
  "createdAt": "2021-08-11T10:00:00Z",
  "updatedAt": "2021-08-11T10:00:00Z"
}
```

### `name`

**Type**: String

**Description**: A human-readable name for the key

**Default value**: `null`

### `description`

**Type**: String

> **Description**: A description for the key

Author: I couldn't find anything for `name` but added a few words for `description`

**Default value**: N/A

### `key`

**Type**: String

**Description**: An alphanumeric key value generated by Meilisearch by hashing the `uid` and the master key on API key creation. Used for authorization when [making calls to a protected Meilisearch instance](/learn/security/master_api_keys.md#communicating-with-a-protected-instance).

This value is also used as the `{key}` path variable to [update](#update-a-key), [delete](#delete-a-key), or [get](#get-one-key) a specific key.

**Default value**: N/A

::: note
Since `key` is a hash of the `uid` and master key, `key` values are deterministic between instances sharing the same configuration. This means if the master key changes, all `key` values are automatically changed.

Since the `key` field depends on the master key, it is computed at runtime and therefore not propagated to dumps and snapshots. As a result, even if a malicious user comes into possession of your dumps or snapshots, they will not have access to your instance's API keys.
:::

### `actions`

**Type**: Array

> **Description**: An array of API actions permitted for the key. `["*"]` for all actions.

Reviewer 2: So, I'm not sure of the best way to put this, but I see we have a repeated warning a few times in the below table:

> "**Non-authorized `indexes` will be omitted from the response**"

I'd like to express this in some form in the description so we could remove it from the table and reduce the overall size/clutter a bit.

```suggestion
**Description**: An array of API actions permitted for the key, represented as strings. API actions are only possible on authorized [`indexes`](#indexes). `["*"]` for all actions.
```

| `settings.update`  | Provides access to the [update settings](/reference/api/settings.md#update-settings) and [reset settings](/reference/api/settings.md#reset-settings) endpoints and equivalents for all subroutes on authorized indexes |
| `stats.get`        | Provides access to the [get stats of an index](/reference/api/stats.md#get-stats-of-an-index) endpoint and the [get stats of all indexes](/reference/api/stats.md#get-stats-of-all-indexes) endpoint. For the latter, **non-authorized `indexes` are omitted from the response** |
| `dumps.create`     | Provides access to the [create dump](/reference/api/dump.md#create-a-dump) endpoint. **Not restricted by `indexes`** |
| `version`          | Provides access to the [get Meilisearch version](/reference/api/version.md#get-version-of-meilisearch) endpoint |
| `keys.get`         | Provides access to the [get all keys](#get-all-keys) endpoint |
| `keys.create`      | Provides access to the [create key](#create-a-key) endpoint |
| `keys.update`      | Provides access to the [update key](#update-a-key) endpoint |

> | `keys.delete`      | Provides access to the [delete key](#delete-a-key) endpoint |

Reviewer 2: If my above suggestion is accepted, I think we can get rid of "on authorized indexes" etc.. `dumps.create` is an exception!

**Default value**: None

### `indexes`

**Type**: Array

> **Description**: An array of indexes the key is authorized to act on. `["*"]` for all indexes. Only the key's [permitted actions](#actions-2) can be used on these indexes

Reviewer 2: ```suggestion
**Description**: An array of indexes the key is authorized to act on. `["*"]` for all indexes. Only the key's [permitted actions](#actions) can be used on these indexes
```

**Default value**: None

### `expiresAt`

**Type**: String

**Description**: Date and time when the key will expire, represented in RFC 3339 format. `null` if the key never expires

**Default value**: None

### `createdAt`

**Type**: String

> **Description**: Date and time when the key was created, represented in RFC 3339 format.

Reviewer 2: ```suggestion
**Description**: Date and time when the key was created, represented in RFC 3339 format
```

**Default value**: `null`

### `updatedAt`

**Type**: String

**Description**: Date and time when the key was last updated, represented in RFC 3339 format

**Default value**: `null`

## Get all keys

<RouteHighlighter method="GET" route="/keys"/>

> Returns a `results` array that lists the 20 most recently created keys. **Expired keys are included in the response**, but deleted keys are not. `total` shows the number of API keys that can be browsed.

Reviewer 2: ```suggestion
Returns the 20 most recently created keys in a `results` array. **Expired keys are included in the response**, but deleted keys are not. The field `total` attached to the response indicates the total number of API keys in existence.
```

### Query parameters

> Results can be paginated by using the `offset` and `limit` query parameters.

Reviewer 2: ```suggestion
Results can be paginated using the `offset` and `limit` query parameters.
```

}
```

For an explanation of these fields, see the [key object](#key-object).

## Create a key

<RouteHighlighter method="POST" route="/keys"/>

Create an API key with the provided description, permissions, and expiration date.

> ### Body

Reviewer 2: I think we should do a table for the body, with links to the extended object definition where necessary. Just seems cleaner, less redundant, and more consistent w/ what we're doing elsewhere. WDYT?

#### `name`

**Type:** String

**Default value:** `null`

A human-readable name for the key.

#### `uid`

**Type:** String

**Default value:** N/A

A [uuid v4](https://www.sohamkamani.com/uuid-versions-explained) to identify the API key. If not specified, it is generated by Meilisearch.

#### `description`

**Type:** String

**Default value:** `null`

An optional description for the key.

> #### `actions` *

Author: >Do we need to specify that the * means required?

I think most users would know as many API references use it, but it might be better to specify it. We could mention it in the overview, so we don't have to mention it on every page

> #### `actions` *

Reviewer 2: Let's assume that users can understand it for now, and if we get a complaint about it we can clarify at that time.

> <RouteHighlighter method="PATCH" route="/keys/{key_or_uid}"/>

Reviewer 2: We are missing a body section + table for this endpoint! It should include `name` and `description`.

Author: I added the links for `name` and `description` in line 287, but a table would be better
